File: src/notification_service/providers.rs
//! Notification providers for different channels

use crate::notification_service::types::{
    NotificationMessage, Recipient, NotificationChannel, DeliveryStatus, 
    DeliveryTracking, ProviderConfig, NotificationResult
};
use async_trait::async_trait;
use chrono::Utc;
use std::collections::HashMap;
use uuid::Uuid;

/// Trait for notification providers
#[async_trait]
pub trait NotificationProvider: Send + Sync {
    /// Send a notification
    async fn send_notification(
        &self,
        message: &NotificationMessage,
        recipient: &Recipient,
    ) -> Result<DeliveryTracking, ProviderError>;

    /// Get provider type
    fn channel(&self) -> NotificationChannel;

    /// Check if provider is healthy
    async fn health_check(&self) -> Result<bool, ProviderError>;

    /// Get provider statistics
    async fn get_stats(&self) -> ProviderStats;
}

/// Email notification provider
pub struct EmailProvider {
    config: ProviderConfig,
    smtp_client: Option<SmtpClient>,
}

impl EmailProvider {
    pub fn new(config: ProviderConfig) -> Result<Self, ProviderError> {
        let smtp_client = if config.enabled {
            Some(SmtpClient::new(&config)?)
        } else {
            None
        };

        Ok(Self { config, smtp_client })
    }
}

#[async_trait]
impl NotificationProvider for EmailProvider {
    async fn send_notification(
        &self,
        message: &NotificationMessage,
        recipient: &Recipient,
    ) -> Result<DeliveryTracking, ProviderError> {
        if !self.config.enabled {
            return Err(ProviderError::ProviderDisabled);
        }

        let email = recipient.email.as_ref()
            .ok_or_else(|| ProviderError::MissingRecipientData("email".to_string()))?;

        let smtp_client = self.smtp_client.as_ref()
            .ok_or_else(|| ProviderError::ProviderNotConfigured)?;

        let tracking_id = Uuid::new_v4().to_string();
        let now = Utc::now();

        // Send email
        let result = smtp_client.send_email(
            email,
            message.subject.as_deref().unwrap_or("Notification"),
            &message.body,
        ).await;

        match result {
            Ok(external_id) => Ok(DeliveryTracking {
                notification_id: message.id.clone(),
                recipient_id: recipient.id.clone(),
                channel: NotificationChannel::Email,
                status: DeliveryStatus::Sent,
                attempts: 1,
                last_attempt: now,
                delivered_at: Some(now),
                error_message: None,
                external_id: Some(external_id),
            }),
            Err(e) => Ok(DeliveryTracking {
                notification_id: message.id.clone(),
                recipient_id: recipient.id.clone(),
                channel: NotificationChannel::Email,
                status: DeliveryStatus::Failed,
                attempts: 1,
                last_attempt: now,
                delivered_at: None,
                error_message: Some(e.to_string()),
                external_id: None,
            }),
        }
    }

    fn channel(&self) -> NotificationChannel {
        NotificationChannel::Email
    }

    async fn health_check(&self) -> Result<bool, ProviderError> {
        if let Some(smtp_client) = &self.smtp_client {
            smtp_client.health_check().await
        } else {
            Ok(false)
        }
    }

    async fn get_stats(&self) -> ProviderStats {
        ProviderStats {
            channel: NotificationChannel::Email,
            total_sent: 0, // Would be tracked in real implementation
            total_failed: 0,
            average_delivery_time_ms: 0,
            last_success: None,
            last_failure: None,
        }
    }
}

/// SMS notification provider
pub struct SMSProvider {
    config: ProviderConfig,
    sms_client: Option<SmsClient>,
}

impl SMSProvider {
    pub fn new(config: ProviderConfig) -> Result<Self, ProviderError> {
        let sms_client = if config.enabled {
            Some(SmsClient::new(&config)?)
        } else {
            None
        };

        Ok(Self { config, sms_client })
    }
}

#[async_trait]
impl NotificationProvider for SMSProvider {
    async fn send_notification(
        &self,
        message: &NotificationMessage,
        recipient: &Recipient,
    ) -> Result<DeliveryTracking, ProviderError> {
        if !self.config.enabled {
            return Err(ProviderError::ProviderDisabled);
        }

        let phone = recipient.phone.as_ref()
            .ok_or_else(|| ProviderError::MissingRecipientData("phone".to_string()))?;

        let sms_client = self.sms_client.as_ref()
            .ok_or_else(|| ProviderError::ProviderNotConfigured)?;

        let tracking_id = Uuid::new_v4().to_string();
        let now = Utc::now();

        // Truncate message for SMS (160 characters typical limit)
        let sms_body = if message.body.len() > 160 {
            format!("{}...", &message.body[..157])
        } else {
            message.body.clone()
        };

        let result = sms_client.send_sms(phone, &sms_body).await;

        match result {
            Ok(external_id) => Ok(DeliveryTracking {
                notification_id: message.id.clone(),
                recipient_id: recipient.id.clone(),
                channel: NotificationChannel::SMS,
                status: DeliveryStatus::Sent,
                attempts: 1,
                last_attempt: now,
                delivered_at: Some(now),
                error_message: None,
                external_id: Some(external_id),
            }),
            Err(e) => Ok(DeliveryTracking {
                notification_id: message.id.clone(),
                recipient_id: recipient.id.clone(),
                channel: NotificationChannel::SMS,
                status: DeliveryStatus::Failed,
                attempts: 1,
                last_attempt: now,
                delivered_at: None,
                error_message: Some(e.to_string()),
                external_id: None,
            }),
        }
    }

    fn channel(&self) -> NotificationChannel {
        NotificationChannel::SMS
    }

    async fn health_check(&self) -> Result<bool, ProviderError> {
        if let Some(sms_client) = &self.sms_client {
            sms_client.health_check().await
        } else {
            Ok(false)
        }
    }

    async fn get_stats(&self) -> ProviderStats {
        ProviderStats {
            channel: NotificationChannel::SMS,
            total_sent: 0,
            total_failed: 0,
            average_delivery_time_ms: 0,
            last_success: None,
            last_failure: None,
        }
    }
}

/// Push notification provider
pub struct PushProvider {
    config: ProviderConfig,
    push_client: Option<PushClient>,
}

impl PushProvider {
    pub fn new(config: ProviderConfig) -> Result<Self, ProviderError> {
        let push_client = if config.enabled {
            Some(PushClient::new(&config)?)
        } else {
            None
        };

        Ok(Self { config, push_client })
    }
}

#[async_trait]
impl NotificationProvider for PushProvider {
    async fn send_notification(
        &self,
        message: &NotificationMessage,
        recipient: &Recipient,
    ) -> Result<DeliveryTracking, ProviderError> {
        if !self.config.enabled {
            return Err(ProviderError::ProviderDisabled);
        }

        let push_client = self.push_client.as_ref()
            .ok_or_else(|| ProviderError::ProviderNotConfigured)?;

        let tracking_id = Uuid::new_v4().to_string();
        let now = Utc::now();

        let mut results = Vec::new();
        
        // Send to all device tokens
        for device_token in &recipient.device_tokens {
            let result = push_client.send_push(
                device_token,
                message.subject.as_deref().unwrap_or("Notification"),
                &message.body,
                &message.data,
            ).await;

            match result {
                Ok(external_id) => results.push(Ok(external_id)),
                Err(e) => results.push(Err(e)),
            }
        }

        // Determine overall status based on results
        let status = if results.iter().any(|r| r.is_ok()) {
            DeliveryStatus::Sent
        } else {
            DeliveryStatus::Failed
        };

        let error_message = if results.iter().all(|r| r.is_err()) {
            Some(results.iter().filter_map(|r| r.as_ref().err())
                .map(|e| e.to_string())
                .collect::<Vec<_>>()
                .join(", "))
        } else {
            None
        };

        let external_id = results.iter().find_map(|r| r.as_ref().ok().cloned());

        Ok(DeliveryTracking {
            notification_id: message.id.clone(),
            recipient_id: recipient.id.clone(),
            channel: NotificationChannel::Push,
            status,
            attempts: 1,
            last_attempt: now,
            delivered_at: if status == DeliveryStatus::Sent { Some(now) } else { None },
            error_message,
            external_id,
        })
    }

    fn channel(&self) -> NotificationChannel {
        NotificationChannel::Push
    }

    async fn health_check(&self) -> Result<bool, ProviderError> {
        if let Some(push_client) = &self.push_client {
            push_client.health_check().await
        } else {
            Ok(false)
        }
    }

    async fn get_stats(&self) -> ProviderStats {
        ProviderStats {
            channel: NotificationChannel::Push,
            total_sent: 0,
            total_failed: 0,
            average_delivery_time_ms: 0,
            last_success: None,
            last_failure: None,
        }
    }
}

/// In-app notification provider
pub struct InAppProvider {
    config: ProviderConfig,
    storage: InAppStorage,
}

impl InAppProvider {
    pub fn new(config: ProviderConfig) -> Result<Self, ProviderError> {
        Ok(Self {
            config,
            storage: InAppStorage::new(),
        })
    }
}

#[async_trait]
impl NotificationProvider for InAppProvider {
    async fn send_notification(
        &self,
        message: &NotificationMessage,
        recipient: &Recipient,
    ) -> Result<DeliveryTracking, ProviderError> {
        if !self.config.enabled {
            return Err(ProviderError::ProviderDisabled);
        }

        let tracking_id = Uuid::new_v4().to_string();
        let now = Utc::now();

        // Store in-app notification
        let notification = InAppNotification {
            id: tracking_id.clone(),
            user_id: recipient.user_id.clone().unwrap_or_else(|| recipient.id.clone()),
            title: message.subject.clone().unwrap_or_else(|| "Notification".to_string()),
            body: message.body.clone(),
            data: message.data.clone(),
            priority: message.priority.clone(),
            read: false,
            created_at: now,
            expires_at: message.scheduled_for,
        };

        let result = self.storage.store_notification(notification).await;

        match result {
            Ok(_) => Ok(DeliveryTracking {
                notification_id: message.id.clone(),
                recipient_id: recipient.id.clone(),
                channel: NotificationChannel::InApp,
                status: DeliveryStatus::Delivered,
                attempts: 1,
                last_attempt: now,
                delivered_at: Some(now),
                error_message: None,
                external_id: Some(tracking_id),
            }),
            Err(e) => Ok(DeliveryTracking {
                notification_id: message.id.clone(),
                recipient_id: recipient.id.clone(),
                channel: NotificationChannel::InApp,
                status: DeliveryStatus::Failed,
                attempts: 1,
                last_attempt: now,
                delivered_at: None,
                error_message: Some(e.to_string()),
                external_id: None,
            }),
        }
    }

    fn channel(&self) -> NotificationChannel {
        NotificationChannel::InApp
    }

    async fn health_check(&self) -> Result<bool, ProviderError> {
        self.storage.health_check().await
    }

    async fn get_stats(&self) -> ProviderStats {
        ProviderStats {
            channel: NotificationChannel::InApp,
            total_sent: 0,
            total_failed: 0,
            average_delivery_time_ms: 0,
            last_success: None,
            last_failure: None,
        }
    }
}

/// Provider statistics
#[derive(Debug, Clone)]
pub struct ProviderStats {
    pub channel: NotificationChannel,
    pub total_sent: u64,
    pub total_failed: u64,
    pub average_delivery_time_ms: u64,
    pub last_success: Option<chrono::DateTime<chrono::Utc>>,
    pub last_failure: Option<chrono::DateTime<chrono::Utc>>,
}

/// Provider errors
#[derive(Debug, thiserror::Error)]
pub enum ProviderError {
    #[error("Provider is disabled")]
    ProviderDisabled,
    
    #[error("Provider not configured")]
    ProviderNotConfigured,
    
    #[error("Missing recipient data: {0}")]
    MissingRecipientData(String),
    
    #[error("Configuration error: {0}")]
    ConfigurationError(String),
    
    #[error("Rate limit exceeded")]
    RateLimitExceeded,
    
    #[error("API error: {0}")]
    ApiError(String),
    
    #[error("Network error: {0}")]
    NetworkError(String),
    
    #[error("Authentication error")]
    AuthenticationError,
    
    #[error("Invalid message format: {0}")]
    InvalidMessageFormat(String),
}

// Mock implementations for client types (would be real implementations in production)
#[derive(Debug)]
struct SmtpClient;

impl SmtpClient {
    fn new(_config: &ProviderConfig) -> Result<Self, ProviderError> {
        // In real implementation, would configure SMTP client
        Ok(Self)
    }
    
    async fn send_email(&self, _to: &str, _subject: &str, _body: &str) -> Result<String, ProviderError> {
        // Mock implementation
        Ok("email_id_123".to_string())
    }
    
    async fn health_check(&self) -> Result<bool, ProviderError> {
        Ok(true)
    }
}

#[derive(Debug)]
struct SmsClient;

impl SmsClient {
    fn new(_config: &ProviderConfig) -> Result<Self, ProviderError> {
        // In real implementation, would configure SMS client (Twilio, etc.)
        Ok(Self)
    }
    
    async fn send_sms(&self, _to: &str, _body: &str) -> Result<String, ProviderError> {
        // Mock implementation
        Ok("sms_id_456".to_string())
    }
    
    async fn health_check(&self) -> Result<bool, ProviderError> {
        Ok(true)
    }
}

#[derive(Debug)]
struct PushClient;

impl PushClient {
    fn new(_config: &ProviderConfig) -> Result<Self, ProviderError> {
        // In real implementation, would configure push client (FCM, APNS, etc.)
        Ok(Self)
    }
    
    async fn send_push(
        &self, 
        _token: &str, 
        _title: &str, 
        _body: &str, 
        _data: &HashMap<String, String>
    ) -> Result<String, ProviderError> {
        // Mock implementation
        Ok("push_id_789".to_string())
    }
    
    async fn health_check(&self) -> Result<bool, ProviderError> {
        Ok(true)
    }
}

#[derive(Debug)]
struct InAppStorage;

impl InAppStorage {
    fn new() -> Self {
        Self
    }
    
    async fn store_notification(&self, _notification: InAppNotification) -> Result<(), ProviderError> {
        // Mock implementation - would store in database
        Ok(())
    }
    
    async fn health_check(&self) -> Result<bool, ProviderError> {
        Ok(true)
    }
}

#[derive(Debug)]
struct InAppNotification {
    id: String,
    user_id: String,
    title: String,
    body: String,
    data: HashMap<String, String>,
    priority: crate::notification_service::types::NotificationPriority,
    read: bool,
    created_at: chrono::DateTime<chrono::Utc>,
    expires_at: Option<chrono::DateTime<chrono::Utc>>,
}

File: src/notification-service/providers.js
//! Notification providers for different channels

const nodemailer = require('nodemailer');
const {
  NotificationChannel,
  DeliveryStatus,
  DeliveryTracking,
  ProviderStats,
  NotificationMessage,
  Recipient
} = require('./types');

class ProviderError extends Error {
  constructor(message, code) {
    super(message);
    this.name = 'ProviderError';
    this.code = code;
  }
}

/**
 * Base provider class
 */
class NotificationProvider {
  constructor(config) {
    this.config = config;
    this.stats = new ProviderStats({
      channel: config.providerType,
      totalSent: 0,
      totalFailed: 0,
      averageDeliveryTimeMs: 0
    });
  }

  /**
   * Send a notification (to be implemented by subclasses)
   */
  async sendNotification(message, recipient) {
    throw new Error('sendNotification must be implemented by subclass');
  }

  /**
   * Get provider type
   */
  channel() {
    return this.config.providerType;
  }

  /**
   * Check if provider is healthy
   */
  async healthCheck() {
    return this.config.enabled;
  }

  /**
   * Get provider statistics
   */
  async getStats() {
    return this.stats;
  }

  /**
   * Update statistics
   */
  updateStats(success, deliveryTime = 0) {
    if (success) {
      this.stats.totalSent++;
      if (deliveryTime > 0) {
        // Update average delivery time
        const totalTime = this.stats.averageDeliveryTimeMs * (this.stats.totalSent - 1) + deliveryTime;
        this.stats.averageDeliveryTimeMs = Math.round(totalTime / this.stats.totalSent);
      }
      this.stats.lastSuccess = new Date();
    } else {
      this.stats.totalFailed++;
      this.stats.lastFailure = new Date();
    }
  }
}

/**
 * Email notification provider
 */
class EmailProvider extends NotificationProvider {
  constructor(config) {
    super(config);
    this.transporter = null;
    this.initializeTransporter();
  }

  /**
   * Initialize SMTP transporter
   */
  initializeTransporter() {
    if (!this.config.enabled) {
      return;
    }

    const smtpConfig = this.config.config;
    if (!smtpConfig.smtp_host || !smtpConfig.username || !smtpConfig.password) {
      console.warn('Email provider not properly configured');
      return;
    }

    this.transporter = nodemailer.createTransporter({
      host: smtpConfig.smtp_host,
      port: parseInt(smtpConfig.smtp_port) || 587,
      secure: smtpConfig.secure === 'true',
      auth: {
        user: smtpConfig.username,
        pass: smtpConfig.password
      }
    });
  }

  /**
   * Send email notification
   */
  async sendNotification(message, recipient) {
    if (!this.config.enabled) {
      throw new ProviderError('Email provider is disabled', 'PROVIDER_DISABLED');
    }

    if (!recipient.email) {
      throw new ProviderError('Recipient email is required', 'MISSING_RECIPIENT_DATA');
    }

    if (!this.transporter) {
      throw new ProviderError('Email transporter not configured', 'PROVIDER_NOT_CONFIGURED');
    }

    const startTime = Date.now();
    const trackingId = `email_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;

    try {
      const mailOptions = {
        from: `${this.config.config.from_name || 'Soroban Security Scanner'} <${this.config.config.from_email || this.config.config.username}>`,
        to: recipient.email,
        subject: message.subject || 'Notification',
        text: message.body,
        html: this.formatAsHtml(message.body)
      };

      const result = await this.transporter.sendMail(mailOptions);
      const deliveryTime = Date.now() - startTime;

      this.updateStats(true, deliveryTime);

      return new DeliveryTracking({
        notificationId: message.id,
        recipientId: recipient.id,
        channel: NotificationChannel.EMAIL,
        status: DeliveryStatus.SENT,
        attempts: 1,
        lastAttempt: new Date(),
        deliveredAt: new Date(),
        externalId: result.messageId
      });
    } catch (error) {
      const deliveryTime = Date.now() - startTime;
      this.updateStats(false, deliveryTime);

      return new DeliveryTracking({
        notificationId: message.id,
        recipientId: recipient.id,
        channel: NotificationChannel.EMAIL,
        status: DeliveryStatus.FAILED,
        attempts: 1,
        lastAttempt: new Date(),
        errorMessage: error.message
      });
    }
  }

  /**
   * Format message as HTML
   */
  formatAsHtml(text) {
    return text
      .replace(/\n/g, '<br>')
      .replace(/\*\*(.*?)\*\*/g, '<strong>$1</strong>')
      .replace(/\*(.*?)\*/g, '<em>$1</em>');
  }

  /**
   * Health check for email provider
   */
  async healthCheck() {
    if (!this.config.enabled || !this.transporter) {
      return false;
    }

    try {
      await this.transporter.verify();
      return true;
    } catch (error) {
      console.error('Email provider health check failed:', error.message);
      return false;
    }
  }
}

/**
 * SMS notification provider (mock implementation)
 */
class SMSProvider extends NotificationProvider {
  constructor(config) {
    super(config);
    this.client = this.initializeClient();
  }

  /**
   * Initialize SMS client (mock)
   */
  initializeClient() {
    if (!this.config.enabled) {
      return null;
    }

    // In a real implementation, would initialize Twilio or other SMS service
    return {
      accountSid: this.config.config.account_sid,
      authToken: this.config.config.auth_token,
      fromNumber: this.config.config.from_number
    };
  }

  /**
   * Send SMS notification
   */
  async sendNotification(message, recipient) {
    if (!this.config.enabled) {
      throw new ProviderError('SMS provider is disabled', 'PROVIDER_DISABLED');
    }

    if (!recipient.phone) {
      throw new ProviderError('Recipient phone is required', 'MISSING_RECIPIENT_DATA');
    }

    if (!this.client) {
      throw new ProviderError('SMS client not configured', 'PROVIDER_NOT_CONFIGURED');
    }

    const startTime = Date.now();
    const trackingId = `sms_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;

    try {
      // Truncate message for SMS (160 characters typical limit)
      const smsBody = message.body.length > 160 
        ? message.body.substring(0, 157) + '...' 
        : message.body;

      // Mock implementation - would use real SMS service
      console.log(`SMS sent to ${recipient.phone}: ${smsBody}`);
      
      const deliveryTime = Date.now() - startTime;
      this.updateStats(true, deliveryTime);

      return new DeliveryTracking({
        notificationId: message.id,
        recipientId: recipient.id,
        channel: NotificationChannel.SMS,
        status: DeliveryStatus.SENT,
        attempts: 1,
        lastAttempt: new Date(),
        deliveredAt: new Date(),
        externalId: trackingId
      });
    } catch (error) {
      const deliveryTime = Date.now() - startTime;
      this.updateStats(false, deliveryTime);

      return new DeliveryTracking({
        notificationId: message.id,
        recipientId: recipient.id,
        channel: NotificationChannel.SMS,
        status: DeliveryStatus.FAILED,
        attempts: 1,
        lastAttempt: new Date(),
        errorMessage: error.message
      });
    }
  }

  /**
   * Health check for SMS provider
   */
  async healthCheck() {
    return this.config.enabled && this.client !== null;
  }
}

/**
 * Push notification provider (mock implementation)
 */
class PushProvider extends NotificationProvider {
  constructor(config) {
    super(config);
    this.fcm = this.initializeFCM();
    this.apns = this.initializeAPNS();
  }

  /**
   * Initialize FCM client (mock)
   */
  initializeFCM() {
    if (!this.config.enabled || !this.config.config.fcm_server_key) {
      return null;
    }

    // In a real implementation, would initialize Firebase Cloud Messaging
    return {
      serverKey: this.config.config.fcm_server_key
    };
  }

  /**
   * Initialize APNS client (mock)
   */
  initializeAPNS() {
    if (!this.config.enabled || !this.config.config.apns_key_id) {
      return null;
    }

    // In a real implementation, would initialize Apple Push Notification Service
    return {
      keyId: this.config.config.apns_key_id,
      teamId: this.config.config.apns_team_id
    };
  }

  /**
   * Send push notification
   */
  async sendNotification(message, recipient) {
    if (!this.config.enabled) {
      throw new ProviderError('Push provider is disabled', 'PROVIDER_DISABLED');
    }

    if (!recipient.deviceTokens || recipient.deviceTokens.length === 0) {
      throw new ProviderError('No device tokens available', 'MISSING_RECIPIENT_DATA');
    }

    const startTime = Date.now();
    const trackingId = `push_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
    let successCount = 0;
    let failureCount = 0;
    const errors = [];

    // Send to all device tokens
    for (const deviceToken of recipient.deviceTokens) {
      try {
        // Mock implementation - would use real push service
        console.log(`Push sent to device ${deviceToken}: ${message.subject}`);
        successCount++;
      } catch (error) {
        failureCount++;
        errors.push(error.message);
      }
    }

    const deliveryTime = Date.now() - startTime;
    const overallSuccess = successCount > 0;

    this.updateStats(overallSuccess, deliveryTime);

    return new DeliveryTracking({
      notificationId: message.id,
      recipientId: recipient.id,
      channel: NotificationChannel.PUSH,
      status: overallSuccess ? DeliveryStatus.SENT : DeliveryStatus.FAILED,
      attempts: 1,
      lastAttempt: new Date(),
      deliveredAt: overallSuccess ? new Date() : null,
      errorMessage: errors.length > 0 ? errors.join(', ') : null,
      externalId: overallSuccess ? trackingId : null
    });
  }

  /**
   * Health check for push provider
   */
  async healthCheck() {
    return this.config.enabled && (this.fcm !== null || this.apns !== null);
  }
}

/**
 * In-app notification provider
 */
class InAppProvider extends NotificationProvider {
  constructor(config) {
    super(config);
    this.storage = new Map(); // In-memory storage for demo
  }

  /**
   * Send in-app notification
   */
  async sendNotification(message, recipient) {
    if (!this.config.enabled) {
      throw new ProviderError('In-app provider is disabled', 'PROVIDER_DISABLED');
    }

    const startTime = Date.now();
    const trackingId = `inapp_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;

    try {
      const notification = {
        id: trackingId,
        userId: recipient.userId || recipient.id,
        title: message.subject || 'Notification',
        body: message.body,
        data: message.data,
        priority: message.priority,
        read: false,
        createdAt: new Date(),
        expiresAt: message.scheduledFor
      };

      // Store in-app notification (in real implementation, would use database)
      const userNotifications = this.storage.get(notification.userId) || [];
      userNotifications.push(notification);
      this.storage.set(notification.userId, userNotifications);

      const deliveryTime = Date.now() - startTime;
      this.updateStats(true, deliveryTime);

      return new DeliveryTracking({
        notificationId: message.id,
        recipientId: recipient.id,
        channel: NotificationChannel.IN_APP,
        status: DeliveryStatus.DELIVERED,
        attempts: 1,
        lastAttempt: new Date(),
        deliveredAt: new Date(),
        externalId: trackingId
      });
    } catch (error) {
      const deliveryTime = Date.now() - startTime;
      this.updateStats(false, deliveryTime);

      return new DeliveryTracking({
        notificationId: message.id,
        recipientId: recipient.id,
        channel: NotificationChannel.IN_APP,
        status: DeliveryStatus.FAILED,
        attempts: 1,
        lastAttempt: new Date(),
        errorMessage: error.message
      });
    }
  }

  /**
   * Get in-app notifications for a user
   */
  getUserNotifications(userId) {
    return this.storage.get(userId) || [];
  }

  /**
   * Mark notification as read
   */
  markAsRead(userId, notificationId) {
    const notifications = this.storage.get(userId) || [];
    const notification = notifications.find(n => n.id === notificationId);
    if (notification) {
      notification.read = true;
      return true;
    }
    return false;
  }

  /**
   * Health check for in-app provider
   */
  async healthCheck() {
    return this.config.enabled;
  }
}

/**
 * Provider factory
 */
class ProviderFactory {
  /**
   * Create provider based on channel
   */
  static createProvider(channel, config) {
    switch (channel) {
      case NotificationChannel.EMAIL:
        return new EmailProvider(config);
      case NotificationChannel.SMS:
        return new SMSProvider(config);
      case NotificationChannel.PUSH:
        return new PushProvider(config);
      case NotificationChannel.IN_APP:
        return new InAppProvider(config);
      default:
        throw new ProviderError(`Unsupported channel: ${channel}`, 'UNSUPPORTED_CHANNEL');
    }
  }

  /**
   * Create default provider configurations
   */
  static getDefaultConfigs() {
    return {
      [NotificationChannel.EMAIL]: {
        providerType: NotificationChannel.EMAIL,
        config: {},
        enabled: false
      },
      [NotificationChannel.SMS]: {
        providerType: NotificationChannel.SMS,
        config: {},
        enabled: false
      },
      [NotificationChannel.PUSH]: {
        providerType: NotificationChannel.PUSH,
        config: {},
        enabled: false
      },
      [NotificationChannel.IN_APP]: {
        providerType: NotificationChannel.IN_APP,
        config: {},
        enabled: true
      }
    };
  }
}

module.exports = {
  NotificationProvider,
  EmailProvider,
  SMSProvider,
  PushProvider,
  InAppProvider,
  ProviderFactory,
  ProviderError
};
